Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/PropertyFilePathCtrl.cpp

```diff
@@ -0,0 +1,324 @@
+        setFocusPolicy(m_browseEdit->lineEdit()->focusPolicy());
+    };
+
+    void PropertyFilePathCtrl::SetFilePath(const AZ::IO::Path& filePath)
+    {
+        m_currentFilePath = filePath;
+        m_browseEdit->setText(filePath.c_str());
+
+        Q_EMIT FilePathChanged();
+    }
+
+    AZ::IO::Path PropertyFilePathCtrl::GetFilePath() const
+    {
+        return m_currentFilePath;
+    }
+
+    void PropertyFilePathCtrl::SetFilter(const QString& filter)
+    {
+        m_filter = filter;
+    }
+
+    void PropertyFilePathCtrl::SetProductExtension(const AZStd::string& extension)
+    {
+        m_productExtension = extension;
+    }
+
+    void PropertyFilePathCtrl::SetDefaultFileName(const AZStd::string& fileName)
+    {
+        m_defaultFileName = fileName;
+    }
+
+    void PropertyFilePathCtrl::OnOpenButtonClicked()
+    {
+        // Retrieve what the pre-selected file path should be for the file dialog
+        QString preselectedFilePath = GetPreselectedFilePath();
+
+        // Popup a native file dialog to choose a new or existing file
+        // Using the AzQtComponents::FileDialog::GetSaveFileName helper will protect the user from entering
+        // a filename with invalid characters for the AP
+        QString newFilePath =
+            AzQtComponents::FileDialog::GetSaveFileName(this, QObject::tr("Open/Save File"), preselectedFilePath, m_filter);
+
+        // Try to set the new file path (if one was selected)
+        // This will handle file paths outside of the safe asset folders
+        HandleNewFilePathSelected(newFilePath);
+    }
+
+    QString PropertyFilePathCtrl::GetPreselectedFilePath() const
+    {
+        QString preselectedFilePath;
+
+        // If we don't have any output file set yet, then generate a default
+        // option for the user
+        if (m_currentFilePath.empty())
+        {
+            AZ::IO::Path defaultPath = AZ::IO::Path(AZ::Utils::GetProjectPath());
+
+            // If a default filename has been specified, append it to
+            // the project path to be the default path
+            if (!m_defaultFileName.empty())
+            {
+                defaultPath /= AZ::IO::Path(m_defaultFileName);
+            }
+
+            preselectedFilePath = defaultPath.c_str();
+        }
+        // Otherwise, we need to find the proper absolute path based on the relative path
+        // that is stored
+        else
+        {
+            // The GetFullSourcePathFromRelativeProductPath API only works on product paths,
+            // so we need to append the product extension to try and find it
+            AZStd::string relativePath = m_currentFilePath.String() + m_productExtension;
+
+            AZStd::string fullPath;
+            bool fullPathIsValid;
+            AssetSystemRequestBus::BroadcastResult(
+                fullPathIsValid, &AssetSystemRequestBus::Events::GetFullSourcePathFromRelativeProductPath, relativePath, fullPath);
```

Comment:
repeat: `fullPathIsValid` must be initialized to `false` as there is no guarantee that the AssetSystemRequestBus has a handler.

Reply:
Good call, fixed